src/components/tl/ucp/allgather/allgather_knomial.c

```diff
@@ -234,6 +262,183 @@ ucc_status_t ucc_tl_ucp_allgather_knomial_start(ucc_coll_task_t *coll_task)
     return ucc_progress_queue_enqueue(UCC_TL_CORE_CTX(team)->pq, &task->super);
 }
 
+void register_memory(ucc_coll_task_t *coll_task){
+
+    ucc_tl_ucp_task_t     *task      = ucc_derived_of(coll_task,
+                                                      ucc_tl_ucp_task_t);
+    ucc_coll_args_t       *args      = &TASK_ARGS(task);
+    ucc_tl_ucp_team_t     *team      = TASK_TEAM(task);
+    ucc_kn_radix_t         radix     = task->allgather_kn.p.radix;
+    uint8_t                node_type = task->allgather_kn.p.node_type;
+    ucc_knomial_pattern_t *p         = &task->allgather_kn.p;
+    void                  *rbuf      = args->dst.info.buffer;
+    ucc_memory_type_t      mem_type  = args->dst.info.mem_type;
+    size_t                 count     = args->dst.info.count;
+    size_t                 dt_size   = ucc_dt_size(args->dst.info.datatype);
+    size_t                 data_size = count * dt_size;
+    ucc_rank_t             size      = task->subset.map.ep_num;
+    ucc_rank_t             broot     = args->coll_type == UCC_COLL_TYPE_BCAST ?
+                                       args->root : 0;
+    ucc_rank_t             rank      = VRANK(task->subset.myrank, broot, size);
+    size_t                 local     = GET_LOCAL_COUNT(args, size, rank);
+    void                  *sbuf;
+    ptrdiff_t              peer_seg_offset, local_seg_offset;
+    ucc_rank_t             peer, peer_dist;
+    ucc_kn_radix_t         loop_step;
+    size_t                 peer_seg_count, local_seg_count;
+    ucc_status_t           status;
+    size_t                 extra_count;
+
+    ucc_tl_ucp_context_t *ctx = UCC_TL_UCP_TEAM_CTX(team);
+    ucp_mem_map_params_t mmap_params;
+    ucp_mem_h            mh;
+    int size_of_list = 1;
+    int count_mh = 0; 
+    ucp_mem_h *mh_list = (ucp_mem_h *)malloc(size_of_list * sizeof(ucp_mem_h));
+
+    mmap_params.field_mask  = UCP_MEM_MAP_PARAM_FIELD_ADDRESS |
+                            UCP_MEM_MAP_PARAM_FIELD_LENGTH  |
+                            UCP_MEM_MAP_PARAM_FIELD_MEMORY_TYPE;
+    mmap_params.memory_type = ucc_memtype_to_ucs[mem_type];
+
+    if (KN_NODE_EXTRA == node_type) {
+        if (p->type != KN_PATTERN_ALLGATHERX) {
+
+            mmap_params.address     = task->allgather_kn.sbuf;
+            mmap_params.length      = local * dt_size;
+            status = ucp_mem_map(ctx->worker.ucp_context, &mmap_params, &mh);
+            if (UCC_OK != status) {
+                task->super.status = status;
+                return;
+            }
+            if (count_mh == size_of_list){
+                size_of_list *= 2;
+                mh_list = (ucp_mem_h *)realloc(mh_list, size_of_list * sizeof(ucp_mem_h));
+            }
+            mh_list[count_mh++] = mh;
```

Comment:
what happended to putting it in function please? or macro?

Reply:
yes I tried but had bug, I'll show you